File: examples/scalajs-react/src/main/scala/todomvc/TodoItem.scala
package todomvc

import japgolly.scalajs.react._
import japgolly.scalajs.react.extra.{Px, Reusability}
import japgolly.scalajs.react.vdom.TagOf
import japgolly.scalajs.react.vdom.html_<^._
import org.scalajs.dom.ext.KeyCode
import org.scalajs.dom.html
import org.scalajs.dom.html.{Input, LI}

object TodoItem {

  case class Props (
    onToggle:        Callback,
    onDelete:        Callback,
    onStartEditing:  Callback,
    onUpdateTitle:   Title => Callback,
    onCancelEditing: Callback,
    todo:            Todo,
    isEditing:       Boolean
  )

  implicit val reusableProps: Reusability[Props] =
    Reusability[Props]((p1, p2) =>
      (p1.todo eq p2.todo) && (p1.isEditing == p2.isEditing)
    )

  case class State(editText: UnfinishedTitle)

  class Backend($: BackendScope[Props, State]) {
    val inputRef: Ref[Input, Input] = Ref[html.Input]

    case class Callbacks(P: Props) {
      val editFieldSubmit: Callback =
        $.state.flatMap(_.editText.validated.fold(P.onDelete)(P.onUpdateTitle))

      val resetText: Callback =
        $.modState(_.copy(editText = P.todo.title.editable))

      val editFieldKeyDown: ReactKeyboardEvent => Option[Callback] =
        e => e.nativeEvent.keyCode match {
          case KeyCode.Escape => Some(resetText >> P.onCancelEditing)
          case KeyCode.Enter  => Some(editFieldSubmit)
          case _              => None
        }
    }

    val cbs: Px[Callbacks] =
      Px.callback($.props).map(Callbacks).withoutReuse.autoRefresh

    val editFieldChanged: ReactEventFromInput => Callback =
      e => {
        /* need to capture event data because React reuses events */
        val captured = e.target.value
        $.modState(_.copy(editText = UnfinishedTitle(captured)))
      }

    def render(P: Props, S: State): TagOf[LI] = {
      val cb = cbs.value()
      <.li(
        ^.classSet(
          "completed" -> P.todo.isCompleted,
          "editing"   -> P.isEditing
        ),
        <.div(
          ^.className := "view",
          <.input(
            ^.className := "toggle",
            ^.`type`    := "checkbox",
            ^.checked   := P.todo.isCompleted,
            ^.onChange --> P.onToggle
          ),
          <.label(
            P.todo.title.value,
            ^.onDoubleClick --> P.onStartEditing
          ),
          <.button(
            ^.className := "destroy",
            ^.onClick --> P.onDelete
          )
        ),
        <.input(
          ^.className   := "edit",
          ^.onBlur     --> cb.editFieldSubmit,
          ^.onChange   ==> editFieldChanged,
          ^.onKeyDown ==>? cb.editFieldKeyDown,
          ^.value       := S.editText.value
        ).withRef(inputRef)
      )
    }
  }

  private val component =
    ScalaComponent.builder[Props]("TodoItem")
      .initialStateFromProps(p => State(p.todo.title.editable))
      .renderBackend[Backend]
      .componentDidUpdate { componentDidUpdate ⇒
        Callback(componentDidUpdate.backend.inputRef.map(_.focus()))
            .when(componentDidUpdate.currentProps.isEditing && !componentDidUpdate.prevProps.isEditing)
            .void
      }
      .build

  def apply(P: Props): VdomElement =
    component.withKey(P.todo.id.id.toString)(P)
}

File: examples/react/src/todoItem.js
import React from 'react';
import classNames from 'classnames';
import {ESCAPE_KEY, ENTER_KEY} from './constants';

class TodoItem extends React.Component {
	constructor(props) {
		super(props);
		this.state = {editText: this.props.todo.title};
		this.textInput = null;
	}

	handleSubmit() {
		let val = !!this.state.editText ? this.state.editText.trim() : null;
		if (val) {
			this.props.onSave(val);
			this.setState({editText: val});
		} else {
			this.props.onDestroy();
		}
	}

	handleEdit() {
		this.props.onEdit();
		this.setState({editText: this.props.todo.title});
	}

	handleKeyDown(event) {
		if (event.which === ESCAPE_KEY) {
			this.setState({editText: this.props.todo.title});
			this.props.onCancel(event);
		} else if (event.which === ENTER_KEY) {
			this.handleSubmit(event);
		}
	}

	handleChange(event) {
		if (this.props.editing) {
			this.setState({editText: event.target.value});
		}
	}

	componentDidUpdate(prevProps) {
		if (!prevProps.editing && this.props.editing && this.textInput) {
			this.textInput.focus();
		}
	}

	render() {
		return (
			<li className={classNames({
				completed: this.props.todo.completed,
				editing: this.props.editing
			})}>
				<div className='view'>
					<input
						className='toggle'
						type='checkbox'
						checked={this.props.todo.completed}
						onChange={() => this.props.onToggle()}
					/>
					<label onDoubleClick={() => this.handleEdit()}>
						{this.props.todo.title}
					</label>
					<button className='destroy' onClick={(e) => this.props.onDestroy(e)}/>
				</div>
				<input
					className='edit'
					ref={input => this.textInput = input}
					value={this.state.editText}
					onBlur={() => this.handleSubmit()}
					onChange={e => this.handleChange(e)}
					onKeyDown={e => this.handleKeyDown(e)}
				/>
			</li>
		);
	}
}

export default TodoItem;
